src/backend/sev/snp/firmware/mod.rs

```diff
@@ -0,0 +1,395 @@
+        }
+    }
+}
+
+impl From<u32> for Indeterminate<Error> {
+    #[inline]
+    fn from(error: u32) -> Indeterminate<Error> {
+        Indeterminate::Known(match error {
+            0 => io::Error::last_os_error().into(),
+            1 => Error::InvalidPlatformState,
+            2 => Error::InvalidGuestState,
+            3 => Error::InvalidConfig,
+            4 => Error::InvalidLen,
+            5 => Error::AlreadyOwned,
+            6 => Error::InvalidCertificate,
+            7 => Error::PolicyFailure,
+            8 => Error::Inactive,
+            9 => Error::InvalidAddress,
+            10 => Error::BadSignature,
+            11 => Error::BadMeasurement,
+            12 => Error::AsidOwned,
+            13 => Error::InvalidAsid,
+            14 => Error::WbinvdRequired,
+            15 => Error::DfFlushRequired,
+            16 => Error::InvalidGuest,
+            17 => Error::InvalidCommand,
+            18 => Error::Active,
+            19 => Error::HardwarePlatform,
+            20 => Error::HardwareUnsafe,
+            21 => Error::Unsupported,
+            22 => Error::InvalidParam,
+            23 => Error::ResourceLimit,
+            24 => Error::SecureDataInvalid,
+            _ => return Indeterminate::Unknown,
+        })
+    }
+}
+
+/// The platform state.
+///
+/// The underlying SEV platform behaves like a state machine and can
+/// only perform certain actions while it is in certain states.
+#[derive(Copy, Clone, Debug, PartialEq)]
+#[non_exhaustive]
+#[repr(u8)]
+pub enum State {
+    /// The platform is uninitialized.
+    Uninitialized,
+
+    /// The platform is initialized, but not currently managing any
+    /// guests.
+    Initialized,
+}
+
+impl std::fmt::Display for State {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        let state = match self {
+            State::Uninitialized => "uninitialized",
+            State::Initialized => "initialized",
+        };
+        write!(f, "{}", state)
+    }
+}
+
+bitflags! {
+    /// Describes the platform state.
+    #[derive(Default)]
+    pub struct Flags: u32 {
+        /// If set, this platform is owned. Otherwise, it is self-owned.
+        const OWNED           = 1 << 0;
+
+        /// If set, encrypted state functionality is present.
+        const ENCRYPTED_STATE = 1 << 8;
+    }
+}
+
+/// The CPU-unique identifier for the platform.
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub struct Identifier(Vec<u8>);
+
+impl From<Identifier> for Vec<u8> {
+    fn from(id: Identifier) -> Vec<u8> {
+        id.0
+    }
+}
+
+impl std::fmt::Display for Identifier {
+    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
+        for b in self.0.iter() {
+            write!(f, "{:02X}", b)?;
+        }
+
+        Ok(())
+    }
+}
+
+/// Information regarding the SEV-SNP platform's TCB version.
+#[derive(Clone, Debug, PartialEq)]
+pub struct SnpTcbStatus {
```

Comment:
Is `Snp` prefix still necessary here (and others), given that we're within `...snp::firmware`?

Reply:
let me fix this